pkg/registry/core/pod/strategy.go

```diff
@@ -128,14 +128,19 @@ func (podStrategy) CheckGracefulDelete(ctx context.Context, obj runtime.Object,
 			period = *pod.Spec.TerminationGracePeriodSeconds
 		}
 	}
-	// if the pod is not scheduled, delete immediately
-	if len(pod.Spec.NodeName) == 0 {
-		period = 0
-	}
-	// if the pod is already terminated, delete immediately
-	if pod.Status.Phase == api.PodFailed || pod.Status.Phase == api.PodSucceeded {
-		period = 0
+	// if the pod has volume resource then don't set the period 0 because it breaks the storage protection feature
+	// for volume types that do not have deletion protection built-in, such as nfs
+	if len(pod.Spec.Volumes) == 0 {
```

Comment:
I think checking for length of volumes won't help, because all Pods are injected with a secret volume by default.

I think we care most about PersistentVolumes (which the StorageProtection feature is for), so maybe we should check if the volume is a PersistentVolumeClaimSource.

Reply:
Fixed. Thanks for the heads up!